### PR TITLE
Experiment with GPO on 2023-02-07

### PR DESCRIPTION
**Pull request by mrakitin**
Works in tandem with https://github.com/NSLS-II/bloptools/pull/5/.

### CODE REVIEW

File: startup/00-startup.py [hunk 1]
```diff
@@ -4,6 +4,13 @@
 from nslsii import configure_base
 from IPython import get_ipython
 
+import sys
+import os
+
+sys.path.insert(0, os.path.expanduser('~/src/bloptools'))
+from bloptools import gp
```
Comment (mrakitin) [inline startup/00-startup.py]: This should be handled via `PYTHONPATH` modification and overlays. Please see https://www.nsls2.bnl.gov/docs/how-to/hot-fix-envs.html#install-a-new-package-for-development for more details.

File: startup/22-shutters.py [hunk 2]
```diff
@@ -0,0 +1,14 @@
...
+class PShutter(Device):
+    open_cmd = Cpt(EpicsSignal, "Cmd:Opn-Cmd")
+    close_cmd = Cpt(EpicsSignal, "Cmd:Cls-Cmd")
+    status = Cpt(EpicsSignalRO, "Pos-Sts")
+    
+psh = PShutter("XF:08BMES-PPS{PSh}", name="psh")
+
+
+# psh.set("Open").wait()
+# psh.set("Close").wait()
```
Comment (mrakitin) [inline startup/22-shutters.py]: Let's combine these commented lines with the commented `#psh = ...` above as they are relevant to this class/object.

File: startup/22-shutters.py [hunk 3]
```diff
@@ -0,0 +1,14 @@
+from nslsii.devices import TwoButtonShutter
+
+#psh = TwoButtonShutter("XF:08BMES-PPS{PSh}", name="psh")
+
+class PShutter(Device):
+    open_cmd = Cpt(EpicsSignal, "Cmd:Opn-Cmd")
+    close_cmd = Cpt(EpicsSignal, "Cmd:Cls-Cmd")
+    status = Cpt(EpicsSignalRO, "Pos-Sts")
```
Comment (mrakitin) [inline startup/22-shutters.py:19]: We can add a custom `set` method to handle the situation of the flaky shutter behavior.

File: startup/92-optimization.py [hunk 4]
```diff
@@ -751,3 +751,14 @@ def move_to_optimized_positions(motors, opt_pos):
...
+# #dofs = [kbh.ush, kbh.dsh]
+dofs = [toroidal_mirror.ush, toroidal_mirror.dsh]
+dofs = [toroidal_mirror.usy, toroidal_mirror.dsy, toroidal_mirror.ush, toroidal_mirror.dsh]
+dofs = [kbv.ush, kbv.dsh, kbh.ush, kbh.dsh]
+
+#dofs = [kbv.ush, kbv.dsh, kbh.ush, kbh.dsh, toroidal_mirror.usy, toroidal_mirror.dsy, toroidal_mirror.ush, toroidal_mirror.dsh]
+
+rel_bounds = {"kbv_ush": [-1e-1, +1e-1], "kbv_dsh": [-1e-1, +1e-1], "kbh_ush": [-1e-1, +1e-1], "kbh_dsh": [-1e-1, +1e-1], "toroidal_mirror_ush": [-1e-1, +1e-1], "toroidal_mirror_dsh": [-1e-1, +1e-1], "toroidal_mirror_usy": [-1e-1, +1e-1], "toroidal_mirror_dsy": [-1e-1, +1e-1]}
+fid_params = {"kbv_ush": -0.0500010, "kbv_dsh": -0.0500010, "kbh_ush": 2.2650053, "kbh_dsh": 3.3120017, "toroidal_mirror_ush": -9.515, "toroidal_mirror_dsh": -3.92, "toroidal_mirror_usy": -6.284, "toroidal_mirror_dsy": -9.2575}
+hard_bounds = np.r_[[fid_params[dof.name] + np.array(rel_bounds[dof.name]) for dof in dofs]].T
```
Comment (mrakitin) [inline startup/92-optimization.py]: I think this module may need an environment variable to optionally enable the optimization-related objects, such as `AUTOALIGNMENT=true` (`yes`, `1`, etc.). It should be turned off by default and can be turned on as needed. Will be reused in `98-suspenders` to disable some of the suspenders interfering with the optimization routines.

File: startup/98-suspenders.py [hunk 5]
```diff
@@ -56,7 +56,7 @@
 RE.install_suspender(ring_suspender)
 #RE.install_suspender(solenoid_v_suspenderCeil)
 #RE.install_suspender(solenoid_v_suspenderFloor)
-RE.install_suspender(I0_suspenderFloor)
+#RE.install_suspender(I0_suspenderFloor)
```
Comment (mrakitin) [inline startup/98-suspenders.py]: Should be optionally turned off via the environment variable `AUTOALIGNMENT` (or a better name).
```suggestion
#RE.install_suspender(I0_suspenderFloor)
```

**Review by mrakitin — COMMENTED** [review]
Here is my preliminary review. We won't merge the PR until after we are done with our experiment on 2/18-19.

**Comment by mrakitin** [comment]
I'll resolve the merge conflict.

**Comment by mrakitin** [comment]
I resolved the conflicts, and the PR should be mergeable now.

**Comment by mrakitin** [comment]
We decided to keep this PR in a Draft state until after we are done with our experiment on 2/18-19. We'll stack the relevant code changes here.

File: startup/00-startup.py (hunk 1 above)
Comment (mrakitin) [reply startup/00-startup.py]: Addressed via https://github.com/NSLS-II-TES/profile_collection/pull/33/commits/787b695091168eec5ab088ea3901622e70eb7476.

File: startup/22-shutters.py (hunk 2 above)
Comment (mrakitin) [reply startup/22-shutters.py]: Addressed via https://github.com/NSLS-II-TES/profile_collection/pull/33/commits/f34b72633d9440531885d4ac083366f81dd253ff.

File: startup/22-shutters.py (hunk 3 above)
Comment (mrakitin) [reply startup/22-shutters.py:19]: To be handled via #34.

File: startup/92-optimization.py [hunk 6]
```diff
@@ -1,70 +1,124 @@
...
-#    - add metadata about the samples
-#    - plot trajectory
+import sys
+import textwrap
+import traceback
+
+from tabulate import tabulate
+
+
+def _extract_tb():
```
Comment (jklynch) [inline startup/92-optimization.py:9]: A short explanation of what this function does would be helpful to the reader.

File: startup/92-optimization.py [hunk 7]
```diff
@@ -1,70 +1,124 @@
...
+        sample_stage.x.name: {"position": sample_stage.x},
+        sample_stage.y.name: {"position": sample_stage.y},
+        # sample_stage.z.name: {'position': sample_stage.z},
+    }
+
+    bound_vals = [(43, 44), (34, 35)]
+    motor_bounds = {}
+    motor_dict_keys = list(motor_dict.keys())
+    for k in range(len(motor_dict_keys)):
+        motor_bounds[motor_dict_keys[k]] = {"position": [bound_vals[k][0], bound_vals[k][1]]}
```
Comment (jklynch) [inline startup/92-optimization.py]: this does not seem to be used

File: startup/92-optimization.py [hunk 8]
```diff
@@ -1,70 +1,124 @@
...
+                headers=("Exception",),
+                maxcolwidths=(term_width - two_borders - indent),
+                **shared_kwargs,
+            ),
+            prefix=" " * indent,
+        )
+    )
+
+
+if auto_alignment_mode():
```
Comment (jklynch) [inline startup/92-optimization.py]: a brief description of what the auto alignment code does would be helpful

**Review by jklynch — CHANGES_REQUESTED** [review]
Only a few minor requests.

File: startup/92-optimization.py (hunk 7 above)
Comment (mrakitin) [reply startup/92-optimization.py]: These parameters are supposed to work with the DE optimization (not currently used). Should we comment these lines out or should we remove them?

Comment (mrakitin) [reply startup/92-optimization.py]: I commented them out in 0704c6e4f8d177bf7ac97667dc48e46b94d182da.

File: startup/92-optimization.py (hunk 8 above)
Comment (mrakitin) [reply startup/92-optimization.py]: I added a comment under that line in 0704c6e4f8d177bf7ac97667dc48e46b94d182da.

File: startup/92-optimization.py (hunk 6 above)
Comment (mrakitin) [reply startup/92-optimization.py:9]: I added a docstring via 0704c6e4f8d177bf7ac97667dc48e46b94d182da.

**Comment by mrakitin** [comment]
@jklynch, are you OK with the changes now?

**Review by thomaswmorris — APPROVED** [review]
Looks good, I don't see any reason why the GPO won't work (but I can't speak to whether or not it affects other things)

**Review by jklynch — APPROVED** [review]
looks good to me!

**Comment by mrakitin** [comment]
Thanks for your reviews! Please feel free to merge.

**Pull request merged**